### PR TITLE
support idle handles for directory

### LINKED ISSUE #2634: Improve performance for readdir when exported via NFS
nfsd will call opendir/releasedir for each `readdir(off)`, and JuiceFS will list call meta.Readdir() for each `opendir`, it will be very slow for large directory (O(N*N)).

We should have a cache for directory, similar to cache of chunks, they will be invalided when mtime of directory is changed.

### PR DESCRIPTION
**Pull request by Hexilee**
Signed-off-by: xixi <hexilee@juicedata.io>

Close #2634

### CODE REVIEW

**Comment by SandyXSD** [comment]
We need a simple test report to know how much this PR improves the situation.
Try `ls` a huge directory (e.g. 1m entries) in nfs client, and compare the time costs.

File: pkg/vfs/handle.go [hunk 1]
```diff
@@ -148,6 +150,29 @@ func (h *handle) Close() {
...
 	v.hanleM.Lock()
 	defer v.hanleM.Unlock()
+	for len(v.idleHandles[inode]) > 0 {
+		length := len(v.idleHandles[inode])
+		h := v.idleHandles[inode][length-1]
+		attr := &meta.Attr{}
+		err := func() syscall.Errno {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+			defer cancel()
+			return v.Meta.GetAttr(meta.WrapContext(ctx), inode, attr)
```
Comment (davies) [inline pkg/vfs/handle.go:160]: this blocks v.hanleM

**Comment by davies** [comment]
We should speedup Meta.Readdir, and cached the result (children) in meta.OpenFiles (same to chunks)

**Review by davies — CHANGES_REQUESTED** [review]
We should speedup Meta.Readdir, and cached the result (children) in meta.OpenFiles (same to chunks)

File: pkg/vfs/handle.go (hunk 1 above)
Comment (Hexilee) [reply pkg/vfs/handle.go]: Fixed

**Pull request closed without merging**